C++ destructors: Ensure "this" is a fully qualified name

tautschnig opened the pull request:
Typechecking does not magically expand `ID_this`, and there isn't sufficient context to use "cpp-this."

Fixes: #661

Only the second commit is new, the first commit is #4470.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
C++ destructors: Ensure "this" is a fully qualified name

Typechecking does not magically expand `ID_this`, and there isn't
sufficient context to use "cpp-this."

Fixes: #661

tautschnig commented:
@kroening @peterschrammel Rebased and ready for review

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c1df4ee31462038ff866070a1d471ab9afd70b44).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106625148

tautschnig committed:
C++ front-end: put "this" parameter in the symbol table

The "this" parameter must carry a unique name for each class and must be
stored in the symbol table. And indeed make sure that all parameters are
in the symbol table, unless the method a member of a class template and
is never used, in which case the method symbol must be removed.
Previously we mostly did this removal, but would fail for methods with a
body not involving any symbols (as we tested for the existence of some
cpp_name).

tautschnig commented:
@kroening @peterschrammel This should now be ready for a review.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 90d565133afa349534e3b12e298570f1fd3fcfb0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107448314
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

kroening commented on src/cpp/cpp_typecheck_function.cpp:
```diff
@@ -171,8 +172,7 @@ irep_idt cpp_typecheckt::function_identifier(const typet &type)
 
-  if(it!=parameters.end() &&
-     it->get_identifier()==ID_this)
+  if(it != parameters.end() && it->get_base_name() == ID_this)
```
Use `get_this()`?

tautschnig replied:
Done, everywhere, in a separate commit.

tautschnig committed:
Use parametert::get_this()

Use the method provided by the parametert API rather than the low-level
approach of a fixed base name.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b18138ced3002e38608945dce0d03bc2dfa19568).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107471346
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
